src/ppx_deriving_yojson.cppo.ml

```diff
@@ -253,6 +253,23 @@ let ser_str_of_record varname labels =
   [%expr let fields = [] in [%e assoc]]
 
 
+let poly_vars_of_type_decl td =
+#if OCAML_VERSION >= (4, 05, 0)
+  let loc = { td.ptype_loc with loc_ghost = true } in
+  Ppx_deriving.fold_left_type_decl (fun acc name -> (mkloc name loc) :: acc) [] td
+#else
+  Ppx_deriving.fold_left_type_decl (fun acc name -> name :: acc) [] td
+#endif
+
+let poly_fun_of_exp_and_type_decl init_exp td =
+#if OCAML_VERSION >= (4, 05, 0)
+  let loc = { td.ptype_loc with loc_ghost = true } in
+  Ppx_deriving.fold_left_type_decl (fun exp name -> Exp.newtype (mkloc name loc) exp)
+    init_exp td
+#else
+   Ppx_deriving.fold_left_type_decl (fun exp name -> Exp.newtype name exp) init_exp td
+#endif
```

Comment:
I think that it would be nice if you could factor the similar parts of the two blocks out of the conditional (for the two functions above). The way I did it in my patch, but there may be a better way, is to define (conditionally) worker functions, so I would guess `mkname` for the first function and `newtype` for the second, and then have the "main" line call these functions: `(fun acc name -> mkname name :: acc)` and `(fun exp name -> newtype name exp)`. I guess the definitions would look somewhat like:

```ocaml
#if OCAML_VERSION >= (4, 05, 0)
  let loc = ... in
  let mkname name = mkloc name loc in
#else
  let mkname name = name in
#endif
```

Reply:
I guess this is where my naive understanding of the AST will come out, but in your suggested approach wouldn't `mkname`'s `loc` depend on the `type_declaration` that we're folding?

In that case, yes, I do think that we should change `fold_left_type_decl` to have a ` 'a -> str -> 'a` function in `4.05.0` and `'a -> string -> 'a` in `< 4.05.0`.